hydra/core/utils.py

```diff
@@ -143,13 +149,20 @@ def run_job(
         ret.overrides = overrides
         # handle output directories here
         Path(str(working_dir)).mkdir(parents=True, exist_ok=True)
-        os.chdir(working_dir)
+
+        if hydra_cfg.hydra.job.chdir:
+            os.chdir(working_dir)
```

Comment:
hmm im not sure I get this - user's application would have already existed after `ret.return_value = task_function(task_cfg)` so I don't think this would interfere with what they do in their application.

Reply:
Here is an example:
```python
# my_app.py
import hydra, os

print(f"1: {os.getcwd()=}")

@hydra.main(config_path=None)
def main(cfg):
    print(f"2: {os.getcwd()=}")
    os.chdir("subdirectory")
    print(f"3: {os.getcwd()=}")

if __name__ == "__main__":
    main()

print(f"4: {os.getcwd()=}")
```
```text
$ python my_app.py hydra.job.chdir=False
1: os.getcwd()='/home/jasha10/hydra.git'
2: os.getcwd()='/home/jasha10/hydra.git'
3: os.getcwd()='/home/jasha10/hydra.git/subdirectory'
4: os.getcwd()='/home/jasha10/hydra.git'
```

Between (3) and (4), the `finally` block from the `utils.py` file gets run, and `os.chdir(old_cwd)` gets called.
The behavior changes if you replace the finally block from the `run_job` function with the below:
```python
    finally:
        HydraConfig.instance().cfg = orig_hydra_cfg
        if _chdir:
            os.chdir(working_dir)
```
Here is the new behavior when you run the example app with the updated `finally` block:
```text
$ python my_app.py hydra.job.chdir=False
1: os.getcwd()='/home/jasha10/hydra.git'
2: os.getcwd()='/home/jasha10/hydra.git'
3: os.getcwd()='/home/jasha10/hydra.git/subdirectory'
4: os.getcwd()='/home/jasha10/hydra.git/subdirectory'
```
I feel that this is the correct behavior because, if the user sets `chdir=False`, then Hydra should not be calling `os.chdir` at all (including in the `finally` block).

Note that this change would require assigning the `_chdir` variable before the start of the `try` block (because otherwise the `_chdir` variable might not exist when the `finally` block gets run).

Reply:
I see - thanks for the example! i've updated the PR. :)